File: src/hunzhu.py
import argparse
import time

from common import const, tool

round_count = 0

JINU_BY_LEVEL = {
    1: const.jinu1,
    2: const.jinu2,
    3: const.jinu3,
}

SELECT_ENEMY_OFFSET = (0, -30)
MAP_WAIT_SECONDS = 40
TURN_LEFT_SECONDS = 0.25
FORWARD_SECONDS = 1.0


def start(round_no, level):
    tool.log(f"第 {round_no} 轮开始[{level} 级激怒]")
    tool.focus_mofa_haqi_window()
    tool.find_and_click(const.fuwenka)
    tool.find_and_click(JINU_BY_LEVEL[level])
    tool.find_and_clickIfExist(const.weijinu, dis=SELECT_ENEMY_OFFSET)
    tool.find_and_click(const.maichongaoyi)
    tool.log(f"第 {round_no} 轮完成")


def parse_args(argv=None):
    levels = sorted(JINU_BY_LEVEL)
    parser = argparse.ArgumentParser(prog="mfhq hunzhu")
    parser.add_argument(
        "level",
        type=int,
        choices=levels,
        nargs="?",
        default=None,
        help="hunzhu level, decides which jinu image to click",
    )
    parser.add_argument(
        "-l",
        "-level",
        "--level",
        dest="level_option",
        type=int,
        choices=levels,
        default=None,
        help="same as the positional level",
    )
    args = parser.parse_args(argv)
    if args.level_option is not None:
        args.level = args.level_option
    elif args.level is None:
        args.level = 1
    return args


def main(argv=None):
    global round_count

    args = parse_args(argv)
    tool.log(f"魂珠脚本启动[{args.level} 级激怒]")
    while True:
        round_count += 1
        start(round_count, args.level)
        if tool.wait_img_appear(const.ditu1, MAP_WAIT_SECONDS) is not None:
            tool.log(f"第 {round_count} 轮：检测到地图，走位")
            tool.press_with_correction('a', TURN_LEFT_SECONDS)
            tool.press_with_correction('w', FORWARD_SECONDS)
        else:
            tool.log(f"第 {round_count} 轮：等待地图超时，跳过走位")
        time.sleep(2)


if __name__ == '__main__':
    main()

File: rust/src/tasks/hunzhu.rs
use std::thread::sleep;
use std::time::Duration;

use crate::assets::{self, Template};
use crate::bot;
use crate::log::log;
use crate::platform::input;

const SELECT_ENEMY_OFFSET: (i32, i32) = (0, -35);
const TURN_LEFT_SECONDS: f64 = 0.25;
const FORWARD_SECONDS: f64 = 1.0;

fn jinu(level: u8) -> &'static Template {
    match level {
        2 => &assets::JINU2,
        3 => &assets::JINU3,
        _ => &assets::JINU1,
    }
}

fn start(round_no: u64, level: u8) {
    log(&format!("第 {round_no} 轮开始[{level} 级激怒]"));
    bot::focus_game();
    bot::click_if_exists(&assets::GUANBI, (0, 0));
    bot::find_and_click(&assets::FUWENKA);
    bot::find_and_click(jinu(level));
    bot::click_if_exists(&assets::WEIJINU, SELECT_ENEMY_OFFSET);
    bot::find_and_click(&assets::MAICHONGAOYI);
    log(&format!("第 {round_no} 轮完成"));
}

pub fn run(level: u8) {
    log(&format!("魂珠脚本启动[{level} 级激怒]"));
    let mut round_count: u64 = 0;
    loop {
        round_count += 1;
        start(round_count, level);
        if bot::wait_appear(&assets::DITU1, bot::DEFAULT_WAIT).is_some() {
            log(&format!("第 {round_count} 轮：检测到地图，走位"));
            input::press_for('a', TURN_LEFT_SECONDS);
            input::press_for('w', FORWARD_SECONDS);
        } else {
            log(&format!("第 {round_count} 轮：等待地图超时，跳过走位"));
        }
        sleep(Duration::from_secs(2));
    }
}
